### PR TITLE
Additional rewriter rules

### PR DESCRIPTION
**Pull request by vhavlena**
This PR adds new rewriting rules. For `SMT-LIB` benchmark it has just a small impact. But it has a significant impact on some symbolic execution benchmarks currently not included to `SMT-LIB`.

### CODE REVIEW

**Comment by vhavlena** [comment]
Results, compared to `devel`
```
# of formulae: 103405
--------------------------------------------------
tool                            ✅    ❌       time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ------  ----  ---------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  102557   848   23151.32   0.23   0.01   1.92  63055    39502        548   269        31        0
z3-noodler-a4d2cbe-84ecd6a  102546   859   25684.72   0.25   0.01   2.41  63057    39489        548   279        32        0
cvc5-1.2.0                   99841  3564   73861.19   0.74   0.01   6.42  60931    38910          2  3553         9        0
z3-4.13.0                    97745  5660  128712.47   1.32   0.01   7.95  58927    38818        211  4914       465       70
--------------------------------------------------
```

image

File: src/smt/theory_str_noodler/expr_cases.cpp [hunk 1]
```diff
@@ -108,4 +108,12 @@ bool is_len_num_leq(expr* e, ast_manager& m, seq_util& m_util_s, arith_util& m_u
...
 
+bool is_indexof_at(expr * index_param, expr* index_str, ast_manager& m, seq_util& m_util_s) {
+    expr *at_str, *at_pos;
+    if(m_util_s.str.is_at(index_param, at_str, at_pos) && index_str->hash() == at_str->hash()) {
+        return true;
+    }
+    return false;
+}
+
 }
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/expr_cases.cpp:119]:
```suggestion
}

```

**Review by jurajsic — APPROVED** [review]
Looks good, just some small things.

Also, can you show the table per benchmark? Were these rules done for some specific benchmark?

**Comment by jurajsic** [comment]
Also, did you run both versions of noodler on the same server/with the same number of parallel runs?

**Comment by vhavlena** [comment]
> Also, did you run both versions of noodler on the same server/with the same number of parallel runs?

Yes, current devel on the current server.

**Comment by vhavlena** [comment]
It has just a minimal effect on `STM-LIB` benchmarks. It has quite a significant effect on `django`, `biopython`, and `thefuck` (I haven't run it now, I am planning to run in on the `not-contains` PR).
```
Benchmark sygus_qgen
# of formulae: 343
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a   343     0    6.32   0.02   0.02   0.01    343        0          0     0         0        0
z3-noodler-a4d2cbe-84ecd6a   343     0    6.50   0.02   0.02   0.01    343        0          0     0         0        0
cvc5-1.2.0                   343     0   80.99   0.24   0.04   0.93    343        0          0     0         0        0
z3-4.13.0                    343     0   10.35   0.03   0.02   0.02    343        0          0     0         0        0
--------------------------------------------------

Benchmark denghang
# of formulae: 999
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a   998     1  129.36   0.13   0.05   0.61    185      813          0     1         0        0
z3-noodler-a4d2cbe-84ecd6a   998     1  130.89   0.13   0.04   0.62    185      813          0     1         0        0
cvc5-1.2.0                   982    17  371.55   0.38   0.00   4.64    170      812          0    17         0        0
z3-4.13.0                    883   116  326.28   0.37   0.08   2.10    186      697          0   105        11        0
--------------------------------------------------

Benchmark automatark
# of formulae: 15995
--------------------------------------------------
tool                           ✅    ❌     time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  -----  ----  -------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  15981    14  3562.87   0.22   0.01   2.75  10468     5513          0    14         0        0
z3-noodler-a4d2cbe-84ecd6a  15982    13  3699.12   0.23   0.01   2.90  10469     5513          0    13         0        0
cvc5-1.2.0                  15903    92  2196.84   0.14   0.00   2.77  10460     5443          0    84         8        0
z3-4.13.0                   15871   124  6107.03   0.38   0.02   3.72  10469     5402          0   123         1        0
--------------------------------------------------

Benchmark stringfuzz
# of formulae: 11618
--------------------------------------------------
tool                           ✅    ❌      time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  -----  ----  --------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  11617     1   2144.49   0.18   0.01   1.11   6355     5262          1     0         0        0
z3-noodler-a4d2cbe-84ecd6a  11616     2   2193.09   0.19   0.01   1.15   6354     5262          1     1         0        0
cvc5-1.2.0                  10946   672  30049.07   2.75   0.00  12.34   6184     4762          0   671         1        0
z3-4.13.0                   11090   528  44638.05   4.03   0.01  14.89   6004     5086          0   403       125        0
--------------------------------------------------

Benchmark redos
# of formulae: 3287
--------------------------------------------------
tool                          ✅    ❌      time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  --------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  3280     7   7514.86   2.29   0.08   4.84   2440      840          0     7         0        0
z3-noodler-a4d2cbe-84ecd6a  3280     7   7657.84   2.33   0.08   4.94   2440      840          0     7         0        0
cvc5-1.2.0                  1116  2171  27007.04  24.20   0.14  34.28    620      496          0  2171         0        0
z3-4.13.0                    888  2399    297.55   0.34   0.01   2.60     48      840          0  2099       300        0
--------------------------------------------------

Benchmark norn
# of formulae: 1027
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  1027     0   16.66   0.02   0.02   0.01    712      315          0     0         0        0
z3-noodler-a4d2cbe-84ecd6a  1027     0   16.18   0.02   0.02   0.01    712      315          0     0         0        0
cvc5-1.2.0                   943    84  247.71   0.26   0.01   2.41    705      238          0    84         0        0
z3-4.13.0                    904   123  278.96   0.31   0.02   5.21    712      192          0   123         0        0
--------------------------------------------------

Benchmark slog
# of formulae: 1976
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  1976     0   51.01   0.03   0.01   0.14    808     1168          0     0         0        0
z3-noodler-a4d2cbe-84ecd6a  1976     0   52.12   0.03   0.01   0.15    808     1168          0     0         0        0
cvc5-1.2.0                  1976     0    1.85   0.00   0.00   0.00    808     1168          0     0         0        0
z3-4.13.0                   1956    20  539.58   0.28   0.01   3.81    788     1168          0    20         0        0
--------------------------------------------------

Benchmark slent
# of formulae: 1128
--------------------------------------------------
tool                          ✅    ❌     time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  -------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  1122     6   118.49   0.11   0.01   1.46    624      498          0     6         0        0
z3-noodler-a4d2cbe-84ecd6a  1121     7    84.30   0.08   0.01   1.14    623      498          0     7         0        0
cvc5-1.2.0                  1104    24  1002.26   0.91   0.00   6.12    618      486          0    24         0        0
z3-4.13.0                   1052    76   345.30   0.33   0.01   4.17    571      481          0    76         0        0
--------------------------------------------------

Benchmark omark
# of formulae: 17
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a     9     8    0.21   0.02   0.01   0.04      2        7          0     7         1        0
z3-noodler-a4d2cbe-84ecd6a     9     8    0.18   0.02   0.01   0.02      2        7          0     7         1        0
cvc5-1.2.0                     5    12    1.03   0.21   0.00   0.46      2        3          0    12         0        0
z3-4.13.0                      7    10    8.06   1.15   0.01   3.01      2        5          0     8         2        0
--------------------------------------------------

Benchmark kepler
# of formulae: 587
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a   584     3   13.21   0.02   0.01   0.06    284      300          0     3         0        0
z3-noodler-a4d2cbe-84ecd6a   584     3   13.24   0.02   0.01   0.06    284      300          0     3         0        0
cvc5-1.2.0                   347   240    0.86   0.00   0.00   0.00     50      297          0   240         0        0
z3-4.13.0                    278   309   25.95   0.09   0.02   0.90    135      143          0   283        26        0
--------------------------------------------------

Benchmark woorpje
# of formulae: 809
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a   771    38  313.57   0.41   0.01   4.03    610      161          0    26        12        0
z3-noodler-a4d2cbe-84ecd6a   773    36  325.99   0.42   0.01   4.68    611      162          0    24        12        0
cvc5-1.2.0                   755    54   47.07   0.06   0.01   0.25    591      164          0    54         0        0
z3-4.13.0                    782    27  295.18   0.38   0.02   4.74    618      164          0    27         0        0
--------------------------------------------------

Benchmark webapp
# of formulae: 681
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a   361   320   39.10   0.11   0.02   0.32    133      228        315     5         0        0
z3-noodler-a4d2cbe-84ecd6a   361   320   39.94   0.11   0.02   0.33    133      228        315     5         0        0
cvc5-1.2.0                   588    93  767.89   1.31   0.02   8.73    165      423          0    93         0        0
z3-4.13.0                    451   230  317.75   0.70   0.02   6.71     37      414        121   109         0        0
--------------------------------------------------

Benchmark kaluza
# of formulae: 19432
--------------------------------------------------
tool                           ✅    ❌     time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  -----  ----  -------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  19386    46   654.64   0.03   0.01   0.71  16277     3109         20    26         0        0
z3-noodler-a4d2cbe-84ecd6a  19372    60   633.71   0.03   0.01   0.98  16275     3097         20    40         0        0
cvc5-1.2.0                  19426     6   459.94   0.02   0.00   0.42  16297     3129          0     6         0        0
z3-4.13.0                   19139   293  1428.86   0.07   0.01   1.71  16022     3117          0   293         0        0
--------------------------------------------------

Benchmark transducer_plus
# of formulae: 91
--------------------------------------------------
tool                          ✅    ❌    time     avg     med     std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  ------  ------  ------  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a     0    91    0.00  nan     nan     nan         0        0         91     0         0        0
z3-noodler-a4d2cbe-84ecd6a     0    91    0.00  nan     nan     nan         0        0         91     0         0        0
cvc5-1.2.0                    42    49  482.14   11.48    3.11   22.06     41        1          0    49         0        0
z3-4.13.0                      1    90    0.01    0.01    0.01  nan         0        1         90     0         0        0
--------------------------------------------------

Benchmark leetcode
# of formulae: 2652
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  2651     1   46.45   0.02   0.01   0.02    867     1784          0     1         0        0
z3-noodler-a4d2cbe-84ecd6a  2649     3   46.90   0.02   0.01   0.02    865     1784          0     0         3        0
cvc5-1.2.0                  2652     0   36.00   0.01   0.00   0.03    867     1785          0     0         0        0
z3-4.13.0                   2652     0   41.89   0.02   0.01   0.11    867     1785          0     0         0        0
--------------------------------------------------

Benchmark str_small_rw
# of formulae: 1880
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  1816    64  259.60   0.14   0.01   1.90      4     1812          4    46        14        0
z3-noodler-a4d2cbe-84ecd6a  1815    65  260.05   0.14   0.01   1.94      4     1811          4    46        15        0
cvc5-1.2.0                  1860    20   34.43   0.02   0.00   0.62     19     1841          2    18         0        0
z3-4.13.0                   1821    59  202.53   0.11   0.01   2.66     25     1796          0    59         0        0
--------------------------------------------------

Benchmark pyex
# of formulae: 23845
--------------------------------------------------
tool                           ✅    ❌      time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  -----  ----  --------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  23748    97   6968.61   0.29   0.17   1.86  20377     3371          0    93         4        0
z3-noodler-a4d2cbe-84ecd6a  23746    99   8642.84   0.36   0.17   3.22  20376     3370          0    99         0        0
cvc5-1.2.0                  23820    25   6002.12   0.25   0.07   2.51  20400     3420          0    25         0        0
z3-4.13.0                   22895   950  55228.54   2.41   0.12  10.38  19624     3271          0   950         0        0
--------------------------------------------------

Benchmark full_str_int
# of formulae: 16968
--------------------------------------------------
tool                           ✅    ❌      time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  -----  ----  --------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a  16817   151   1311.26   0.08   0.01   1.46   2496    14321        117    34         0        0
z3-noodler-a4d2cbe-84ecd6a  16824   144   1881.17   0.11   0.01   1.87   2503    14321        117    26         1        0
cvc5-1.2.0                  16963     5   5072.37   0.30   0.01   1.39   2521    14442          0     5         0        0
z3-4.13.0                   16732   236  18620.60   1.11   0.01   6.38   2476    14256          0   236         0        0
--------------------------------------------------

Benchmark snia
# of formulae: 70
--------------------------------------------------
tool                          ✅    ❌    time     avg     med     std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  ------  ------  ------  -----  -------  ---------  ----  --------  -------
z3-noodler-785a5cb-84ecd6a    70     0    0.61    0.01    0.01    0.01     70        0          0     0         0        0
z3-noodler-a4d2cbe-84ecd6a    70     0    0.66    0.01    0.01    0.01     70        0          0     0         0        0
cvc5-1.2.0                    70     0    0.03    0.00    0.00    0.00     70        0          0     0         0        0
z3-4.13.0                      0    70    0.00  nan     nan     nan         0        0          0     0         0       70
--------------------------------------------------
```

**Comment by jurajsic** [comment]
Looks good, you can merge.

**Pull request merged**